src/graphnet/data/extractors/utilities/types.py

```diff
@@ -0,0 +1,243 @@
+"""Utility methods for checking the types of objects."""
+
+import inspect
+from typing import Any, Dict, List, Optional
+
+from graphnet.data.extractors.utilities.collections import (
+    transpose_list_of_dicts,
+    serialise,
+    flatten_nested_dictionary,
+)
+from graphnet.data.extractors.utilities.frames import (
+    get_om_keys_and_pulseseries,
+)
+from graphnet.utilities.logging import get_logger
+
+logger = get_logger()
+
+try:
+    from icecube import (
+        icetray,
+    )  # pyright: reportMissingImports=false
+except ImportError:
+    logger.warning("icecube package not available.")
+
+
+def is_boost_enum(obj: Any) -> bool:
+    """Check whether `obj` inherits from Boost.Python.enum."""
+    for cls in type(obj).__bases__:
+        if "Boost.Python.enum" in str(cls):
+            return True
+    return False
+
+
+def is_boost_class(obj: Any) -> bool:
+    return "Boost.Python.class" in str(type(obj))
+
+
+def is_icecube_class(obj: Any) -> bool:
+    """Check whether `obj` is an IceCube-specific class."""
+    classname = str(type(obj))
+    return "icecube." in classname
+
+
+def is_type(obj: Any) -> bool:
+    """Checks whether `obj` is a type, and not an instance."""
+    return type(obj).__name__ == "type"
+
+
+def is_method(obj: Any) -> bool:
+    """Check whether `obj` is a method."""
+    return inspect.ismethod(obj) or "Boost.Python.function" in str(type(obj))
+
+
+def get_member_variables(
+    obj: Any, return_discarded: bool = False
+) -> List[str]:
+    """Returns list of valid member variables.
+
+    Ignoring mangled (__*) variables, types, methods, and Boost enums.
+    """
+    valid_member_variables = []
+    discarded_member_variables = {
+        "mangled": [],
+        "is_type": [],
+        "invalid_attr": [],
+        "is_method": [],
+        "is_boost_enum": [],
+        "is_boost_class": [],
+    }
+    for attr in dir(obj):
+        if attr.startswith("__"):
+            discarded_member_variables["mangled"].append(attr)
+            continue
+
+        try:
+            value = getattr(obj, attr)
+        except RuntimeError:
+            discarded_member_variables["invalid_attr"].append(attr)
+            continue
+
+        if is_type(value):
+            discarded_member_variables["is_type"].append(attr)
+        elif is_method(value):
+            discarded_member_variables["is_method"].append(attr)
+        elif is_boost_enum(value):
+            discarded_member_variables["is_boost_enum"].append(attr)
+        elif is_boost_class(value):
+            discarded_member_variables["is_boost_class"].append(attr)
+        else:
+            valid_member_variables.append(attr)
+
+    if return_discarded:
+        return valid_member_variables, discarded_member_variables
+
+    return valid_member_variables
+
+
+def cast_object_to_pure_python(obj: Any) -> Any:
+    """Extracts all possible pure-python data from `obj`."""
+
+    logger.debug(f"Value: {obj}")
```

Comment:
Could we expand on this doc string a bit? Just to make the recursive nature of this function a bit clearer

Reply:
(Seems to be very central to the extractor, so putting in a nice fat doc string here is probably worthwhile)